accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,191 @@
+
+Generated code should not be generated in System namespaces, it should prefer the user's namespace.  Generated code should be triple-slashed commented and be free from compiler warnings.  Generated code should avoid name collisions with user code and provide affordances for resolving name conflicts.  As much as possible generated code should meet the quality standards of .NET library code, using nullable annotations, meeting framework design guidelines, naming guidelines, etc.  It should feel like a natural extension of the library.
+
+### Deterministic failure
+
+The absence of a required source generator from code that requires it should produce a clear diagnosable error message from the compiler, like the error from a missing reference.  A codebase that was written to make use of a source-generator should never compile and silently omit the source generated code.  For example: if a source generator uses an attribute to activate the generator and then generates extension methods with overloads that are automatically preferred, the source generator should define the attribute in the source generator, not the framework library.  This will ensure that the project fails to compile if the source generator is absent.
+
+### Source generator servicing
+
+Source generators contribute significant code to user assemblies and that code may have issues.  We have very little control over changing customer assemblies once they have consumed a source generator.  We cannot and should not attempt to change the behavior of customer's assemblies (eg: hot patch) to fix issues.  Should an issue be found we should fix it in the source generator and ship a new version, just as we would if we found an issue with compiler.  Customers desiring this fix must proactively acquire the new code-generator, typically through an updated SDK which carries the ref pack, rebuild their code, and update their production applications.
+
+Due to the limited servicing agility for code-generated code, we should be very careful about the complexity of the code which is generated.  Where possible we should limit generated code to "glue code", only that which must be specific to the user's assembly or types.  We should try to put more complex code inside the framework or library itself, even if it means exposing public API that is specifically for source generated code to call.
+
+Inbox source generators will be serviced in their framework-specific location by shipping a new version of the ref-pack.  Should a bug exist in a source generator, we will evaluate the bug against the bar for each framework in which it exists and patch them independently.  In this way the servicing process for these source generators will match the .NET runtime.
```

Comment:
This worries me. The servicing bar for the .NET runtime is very high compared to the .NET SDK which is very low. Generators exist more at the SDK layer and today SDK level servicing is relatively common. If we are putting Framework source generators where they require the .NET runtime servicing bar then I feel like we're going to be potentially setting ourselves up for pain here.

Reply:
The servicing bar for the compiler is even higher than the runtime though, right?
It's the same set of folks enforcing the bar and they can decide how to apply it.  Servicing changes for the SDK still need to go to tactics.  These components are emitting code into the user's assembly that impacts their application at runtime so I think it's correct to apply a higher bar than design-time-only SDK code.

Reply:
> The servicing bar for the compiler is even higher than the runtime though, right?

No it's much lower.

When the runtime is serviced it is forced onto customers. For a given .NET release there is one runtime for the X different SDKs that we will produce on top of it. The compiler ships inside all these SDKs and can be free to be aggressive with changes because customers have to proactively consume the SDK. 

> Servicing changes for the SDK still need to go to tactics

There are two aspects to consider here:
1. How often the SDK changes for a given .NET runtime release 
2. The bar for servicing the SDK 

I'm talking about both here but primarily (1). The SDK changes *frequently* and with major revisions for a given release. It's not uncommon for this to reveal bugs / issues in existing analyzers / generators that need to be patched. Why do the runtime generators think that they won't have the same type of issues that we've seen with other components historically?

> These components are emitting code into the user's assembly that impacts their application at runtime so I think it's correct to apply a higher bar than design-time-only SDK code.

We have a fairly established bar for code changes in customer code. This feels like you're saying that existing bar (which we've used for many many years) is wrong. So I'd like to get more elaboration on this.

Reply:
> why do the runtime generators think that they won't have the same type of issues that we've seen with other components historically?

I don't think I was implying we wouldn't have to service these components or that they wouldn't have issues.  I'm imagining the majority of our servicing fixes for the generator are related to the generator's functionality, not it's interactions with the platform.  The source-generator platform is an app model where it needs to continue to work version-to-version, and the compiler would be just as eager to fix any compatibility bugs it introduced to break a source generator as it would be to tell all the source generators in the ecosystem to change to adopt a breaking change.  For the bad bugs we can always go back and service the generator in a particular framework's servicing branch, but for some it won't meet the bar and the "bug" would go away when the framework goes out of support.

>  This feels like you're saying that existing bar (which we've used for many many years) is wrong. So I'd like to get more elaboration on this.

I don't think I made any claims that the existing bar was wrong.  I'm saying a higher bar than **design-time only code**.  My impression is that generators should have a bar closer to framework code than the compiler due to the amount of loosely specified complexity they emit: small amount of user code results in large amount of complexity in the user's assembly.  I agree they are different than inbox framework code since they require SDK update and recompile, but probably require more scrutiny than OOB nuget packages (which also require a design time update) since the updates are bundled in the SDK as part of the default experience.  I tried to cover some of the differences here https://github.com/dotnet/designs/blob/e69303f5023302d6e60ac2220a0ff2d819b69609/accepted/2021/InboxSourceGenerators.md#inplace-servicing-updates-are-good-enough-for-the-compiler-why-not-source-generators

Another thing to consider that's related to this: most of the generators we're looking at now also ship with the libraries they extend in NuGet packages.  So we still need to drive fixes through dotnet/runtime to get those packages updated even if the generator also shipped framework-independent in the SDK.